lib/penelope/ml/word2vec/index.ex

```diff
@@ -206,28 +218,50 @@ defmodule Penelope.ML.Word2vec.Index do
   @doc """
   inserts a word vector tuple into a word2vec index
   """
-  @spec insert!(index::Index.t, record::{String.t, Vector.t}) :: :ok
-  def insert!(%Index{vector_size: vector_size} = index,
-              {_term, vector} = record) do
+  @spec insert!(index::Index.t, record::{String.t, pos_integer, Vector.t})
+    :: :ok
+  def insert!(%Index{vector_size: vector_size, header: header} = index,
+              {term, id, vector} = record) do
     actual_size = div(byte_size(vector), 4)
     unless actual_size === vector_size do
       raise IndexError, "invalid vector size: #{actual_size} != #{vector_size}"
     end
-    case index
-         |> get_table(elem(record, 0))
-         |> :dets.insert(record) do
-      :ok              -> :ok
+    table = get_table(index, elem(record, 0))
+
+    with :ok <- :dets.insert(header, {id, term}),
+         :ok <- :dets.insert(table, record) do
+      :ok
+    else
+      {:error, reason} -> raise IndexError, inspect(reason)
+    end
+  end
+
+  @doc """
+  retrieves a term by its id
+
+  if found, returns the term string
+  otherwise, returns nil
+  """
+  @spec fetch!(index::Index.t, id::pos_integer) :: String.t | nil
+  def fetch!(%{name: name} = index, id) do
+    :e2qc.cache(name, id, fn -> do_fetch(index, id) end)
+  end
+
+  def do_fetch(%{header: header}, id) do
```

Comment:
Do you mean `do_fetch` and `do_lookup` to be private? Noticed that there wasn't any doc/spec.